internal/internal_worker_cache.go

```diff
@@ -0,0 +1,124 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package internal
+
+import (
+	"runtime"
+	"sync"
+
+	"go.uber.org/atomic"
+
+	"go.temporal.io/sdk/internal/common/cache"
+)
+
+// A shared cache workers can use to store workflow state. The cache is expected to be initialized with the first worker
+// to be instantiated, and freed with the last worker to terminate. IE: All workers have a pointer to it.
+type workerCache struct {
+	workflowCache *cache.Cache
+}
+
+// Don't touch this except via methods on workerCache. Holds shared workflow cache.
+var workflowCache cache.Cache
+var stickyCacheSize = defaultStickyCacheSize
+var stickyCacheLock sync.Mutex
+
+// Holds the number of outstanding references to the cache so we can zero it out if all workers are dead
+var workerRefcount atomic.Int32
```

Comment:
It is confusing to have both mutex and atomic in the same functions. IMHO just using mutex and int counter is simpler and perf is not an issue in this code path.